lightning/src/offers/merkle.rs

```diff
@@ -76,13 +76,34 @@ impl AsRef<TaggedHash> for TaggedHash {
 
 /// Error when signing messages.
 #[derive(Debug, PartialEq)]
-pub enum SignError<E> {
+pub enum SignError<E = ()> {
 	/// User-defined error when signing the message.
 	Signing(E),
 	/// Error when verifying the produced signature using the given pubkey.
 	Verification(secp256k1::Error),
 }
 
+/// A function for signing a [`TaggedHash`].
+pub trait SignFunction<T: AsRef<TaggedHash>> {
```

Comment:
> A trait in the C bindings cannot be generic (its always concretized into one instance), but generally I'm not sure this is quite the right interface anyway - shouldn't an implementation of this interface have some idea of what its signing aside from just the `TaggedHash`?

The generic version is primarily so it can be used by the module-visible `sign_message` function for different types. However, the public interfaces use concrete types (e.g., `UnsignedBolt12Invoice::sign` takes `SignFunction<UnsignedBolt12Invoice>`).

Assuming type aliases aren't an option, we could add parameter-specific traits that have blanket implementations for `Fn` and implementations of `SignFunction` for their concrete types.

```rust
trait SignBolt12InvoiceFn {
       type Error;

       fn sign_invoice(&self, message: &UnsignedBolt12Invoice) -> Result<Signature, Self::Error>;
}

impl<F, E> SignBolt12InvoiceFn for F
where
       F: Fn(&UnsignedBolt12Invoice) -> Result<Signature, E>,
{
       type Error = E;

       fn sign_invoice(&self, message: &UnsignedBolt12Invoice) -> Result<Signature, E> {
               self(message)
       }
}

impl<F: SignBolt12InvoiceFn> SignFunction<UnsignedBolt12Invoice> for F {
       type Error = <Self as SignFunction<UnsignedBolt12Invoice>>::Error;

       fn sign(&self, message: &UnsignedBolt12Invoice) -> Result<Signature, Self::Error> {
               self.sign_invoice(message)
       }
}

```

> (or, if we think it doesn't need context, let's just pass it the `TaggedHash` directly?)

I thought there may have been a need by VLS but can't recall the specifics. Still would be nice to have the actual message.

Reply:
> > The generic version is primarily so it can be used by the module-visible sign_message function for different types. However, the public interfaces use concrete types (e.g., UnsignedBolt12Invoice::sign takes SignFunction).
> 
> Mmm, right, still its a bit awkward for users who want to verify, you have to have two separate implementations (on the same struct?) rather than a single one. I guess its not that bad, but its still a bit more awkward than just two separate functions.
> 

Not sure I follow the point about verifying.

> I'm not convinced all this work is worth it just to verify the signature. We could simply skip that step, remove `sign_message` and move on?

Eh, was just about done implementing it. So pushed the version with dedicated traits. Required a small change from what I wrote above, though just simplifying it.

Plus, at least it verifies that they signed the right thing now that `TaggedHash` is no longer part of any of the signing interfaces. 😛